### PR TITLE
Comestibles migration (part 1) [RDY]

### PR DESCRIPTION
**Pull request by mugling**
- Adds support for `MIGRATION` (see examples in `migrations.json`). Future PR's will build upon this functionality to deal with a lot of our legacy (mis)naming issues.
- Adds `IRRADIATED` flag and migrates the old `irradiated_foo` items to use it
- Adds new `fruit` and `vegetable` item groups and their irradiated equivalents
- Implements some initial `vitamin` groupings. Notably `vegetable_leafy` is now an acceptable source of `calcium`
- Part 2 will address item spawns for irradiated fruit and vegetables (which are a valuable commodity now we have a vitamins system)

### CODE REVIEW

**Comment by mugling** [comment]
Ready for review, apologies for size but there's no logically smaller unit for dropping all of the `irradiated_foo` items from the recipes.

File: data/json/itemgroups/food.json [hunk 1]
```diff
@@ -0,0 +1,63 @@
+[
+  {
+    "id": "fruit",
+    "type": "item_group",
+    "//": "Commonn fresh fruits excluding types only available via foraging or farming",
```
Comment (illi-kun) [inline data/json/itemgroups/food.json]: typo (```Commonn```).

**Comment by kevingranade** [comment]
> Part 2 will address item spawns for irradiated fruit and vegetables (which are a valuable commodity now we have a vitamins system)

So this PR prevents irradiated fruit from spawning anywhere? That's not really ok.
Also when I loaded a save it didn't convert the irradiated fruit I came across, I ended up with debug messages and "undefined-irradiated_foo"

File: data/json/items/comestibles/fruit.json [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+  },
+  {
+    "id": "blackberries",
+    "copy-from": "fruit_berry",
+    "type": "COMESTIBLE",
+    "name": "blackberry",
+    "name_plural": "blackberries",
+    "description": "A darker cousin of raspberry.",
+    "weight": 226,
+    "volume": 1,
```
Comment (Coolthulhu) [inline data/json/items/comestibles/fruit.json]: 1 volume per each berry?

File: data/json/items/comestibles/fruit.json [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+    "fun": 3
+  },
+  {
+    "id": "coconut",
+    "copy-from": "fruit",
+    "type": "COMESTIBLE",
+    "name": "coconut",
+    "description": "A fruit with a hard and hairy shell.",
+    "weight": 1300,
+    "volume": 1,
```
Comment (Coolthulhu) [inline data/json/items/comestibles/fruit.json]: Why explicitly specify 1 volume for everything?

File: data/json/items/comestibles/vegetable.json [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    "name": "broccoli",
+    "name_plural": "broccoli",
+    "description": "It's a bit tough, but quite delicious.",
+    "weight": 170,
+    "volume": 1,
+    "price": 400,
+    "spoils_in": 160,
+    "nutrition": 30,
+    "healthy": 2,
+    "flags": [ "EATEN_HOT" ]
```
Comment (Coolthulhu) [inline data/json/items/comestibles/vegetable.json]: Was this in before? Heating up a broccoli over a hotplate isn't exactly the proper way to cook it

File: data/json/items/comestibles/vegetable.json [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+    "copy-from": "vegetable",
+    "type": "COMESTIBLE",
+    "name": "celery",
+    "description": "Neither tasty nor very nutritious, but it goes well with salad.",
+    "weight": 80,
+    "volume": 1,
+    "price": 200,
+    "spoils_in": 150,
+    "nutrition": 5,
+    "quench": 3
```
Comment (Coolthulhu) [inline data/json/items/comestibles/vegetable.json]: No health or extra vitamins?

File: data/json/items/comestibles/vegetable.json [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    "weight": 150,
+    "volume": 1,
+    "price": 300,
+    "spoils_in": 150,
+    "nutrition": 11,
+    "quench": 7,
+    "healthy": 1
+  },
+  {
+    "id": "garlic",
```
Comment (Coolthulhu) [inline data/json/items/comestibles/vegetable.json]: Note: there are two garlics - one is just 1/6 of the other. It would be good to merge them into one item (preferably the more granular one, since no sane person will eat a whole bulb of garlic in one go).

File: data/json/items/comestibles/vegetable.json [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+    "spoils_in": 48,
+    "nutrition": 14,
+    "healthy": 1,
+    "fun": -3
+  },
+  {
+    "id": "mushroom_magic",
+    "copy-from": "mushroom",
+    "type": "COMESTIBLE",
+    "extend": { "flags" : "HIDDEN_HALLU" }
```
Comment (Coolthulhu) [inline data/json/items/comestibles/vegetable.json]: Note: HIDDEN_HALLU/POISON themselves do NOT cause poison or hallucinations. iuse functions do that.

File: data/json/items/comestibles/vegetable.json [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+    "name": "sugar beet",
+    "description": "This fleshy root is ripe and flowing with sugars; just takes some processing to extract them.",
+    "weight": 310,
+    "volume": 4,
+    "price": 50,
+    "charges": 4,
+    "spoils_in": 2280,
+    "nutrition": 5,
+    "fun": -10,
+    "stim": 1
```
Comment (Coolthulhu) [inline data/json/items/comestibles/vegetable.json]: That stim is probably from when sugar was a stimulant. But sugar is no longer a stimulant since that was based on a myth. Unless sugar beets contain psychoactive drugs, they shouldn't have stim.

File: data/json/items/comestibles/vegetable.json [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+    "name_plural": "wild vegetables",
+    "description": "An assortment of edible-looking wild plants.  Most are quite bitter-tasting.",
+    "weight": 238,
+    "volume": 1,
+    "price": 60,
+    "spoils_in": 80,
+    "nutrition": 20,
+    "healthy": 1,
+    "fun": -10
+  },
```
Comment (Coolthulhu) [inline data/json/items/comestibles/vegetable.json]: Note: vild veggies had parasite chance. Now they don't.

File: data/json/items/comestibles/vegetable.json [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+  {
+    "id": "zucchini",
+    "copy-from": "vegetable",
+    "type": "COMESTIBLE",
+    "name": "zucchini",
+    "description": "A tasty summer squash.",
```
Comment (Coolthulhu) [inline data/json/items/comestibles/vegetable.json]: Description says "tasty", statblock disagrees.

File: doc/JSON_FLAGS.md [hunk 11]
```diff
@@ -747,6 +747,7 @@ Some armor flags, such as `WATCH` and `ALARMCLOCK` are compatible with other ite
 
 - ```EATEN_HOT``` Morale bonus for eating hot.
 - ```EATEN_COLD``` Morale bonus for eating cold.
+- ```IRRADIATED``` Never spoils
```
Comment (Coolthulhu) [inline doc/JSON_FLAGS.md]: The whole idea of having irradiated as a flag is a bit of a hack. The most proper way to handle this would be to have a "sealed bag" container which would preserve contents, but become unsealed on access.

Will suffice for now, assuming it is handled correctly.

File: data/json/items/comestibles/vegetable.json (hunk 10 above)
Comment (mugling) [inline data/json/items/comestibles/vegetable.json]: It's a copy-paste from the original but agreed, I'll rework that

File: data/json/items/comestibles/vegetable.json (hunk 9 above)
Comment (mugling) [inline data/json/items/comestibles/vegetable.json]: That was lost a little while back. IRL the risks are exceptionally low (to nil). I'd like to rework a better parasitism system rather than bolt the flawed one back on.

File: data/json/items/comestibles/vegetable.json (hunk 8 above)
Comment (mugling) [inline data/json/items/comestibles/vegetable.json]: Fair enough

File: data/json/items/comestibles/vegetable.json (hunk 7 above)
Comment (mugling) [inline data/json/items/comestibles/vegetable.json]: Perhaps the eating code should be updated to handle the flags?

Comment (Coolthulhu) [inline data/json/items/comestibles/vegetable.json]: If you want to go that route, then it will also require updating NPC eating code to look for those flags.

Comment (mugling) [inline data/json/items/comestibles/vegetable.json]: I'm not averse to either route although I suspect the flags only approach is a cleaner solution?

Comment (Coolthulhu) [inline data/json/items/comestibles/vegetable.json]: Could be, since then we would have all consumption-related code in one place, easy to look up from the code side, easy to pass on further (once crafting system supports that) and we could separate consumption from use greatly tidying apply and eat menus, but it would also mean a loss of a certain kind of customization.

For example, alcohol, caffeine, royal jelly, tainted tornado, mutant body parts, marloss - all of those would need special handling.

It could be a big deal, so it would be better to make a whole issue to discuss the change.
But if we decide not to go the whole way in and remove comestible iuses, it would be better to have poison and hallucination as iuse. And parasites too.

Comment (mugling) [inline data/json/items/comestibles/vegetable.json]: Went with the simpler solution in 97d117d for now

File: doc/JSON_FLAGS.md (hunk 11 above)
Comment (mugling) [inline doc/JSON_FLAGS.md]: Ok, this works well as incremental progress towards that goal

File: data/json/items/comestibles/vegetable.json (hunk 6 above)
Comment (mugling) [inline data/json/items/comestibles/vegetable.json]: Fixed in e252eb8

File: src/item.cpp [hunk 12]
```diff
@@ -2755,7 +2768,7 @@ std::set<matec_id> item::get_techniques() const
 
 bool item::goes_bad() const
 {
-    return is_food() && type->comestible->spoils;
+    return is_food() && !has_flag( "IRRADIATED" ) && type->comestible->spoils;
```
Comment (Coolthulhu) [inline src/item.cpp]: I didn't check how often is `goes_bad` called. It might be worth looking at, since it is possible that it gets called a lot.

In such a case it would probably be a good idea to cache spoil protection as a `bool` or at least getting the "IRRADIATED" string out into a static constant.

**Comment by mugling** [comment]
> So this PR prevents irradiated fruit from spawning anywhere?

Anyone have any suggestions as to spawn locations?

> Also when I loaded a save it didn't convert the irradiated fruit I came across, I ended up with debug messages and "undefined-irradiated_foo"

Struggling to reproduce. Can you try after the recent rebase against `master` and see if the issue persists?

**Comment by mugling** [comment]
I've started reworking food item groups in a separate yet-to-push PR which will include irradiated food. This PR is huge and I want to avoid extending it further for now

**Comment by kevingranade** [comment]
> This PR is huge and I want to avoid extending it further for now

That's because you're doing it in the wrong order, get the migration feature in with a minimal set of use cases to validate the code (that's only a few hundred lines of code in item_factory and savegame_json, and a miniminal amount of json changes), then start tackling the actual migrations.  It's simply not ok to disable spawning of an entire class of items to refactor them.

> apologies for size but there's no logically smaller unit for dropping all of the irradiated_foo items from the recipes.

That's a very strong sign that it should have a dedicated PR instead of being mixed into a PR with a bunch of other feature additions and refactorings.

**Comment by mugling** [comment]
There isn't a smaller unit though? We don't want half of the irradiated fruit using the old system and half using the new?

**Comment by Coolthulhu** [comment]
> We don't want half of the irradiated fruit using the old system and half using the new?

Partial migrations are acceptable, when doable and too huge to merge all at once. Say, first add the new mechanics and itemgroups, then do the migration with removed items being marked as obsolete, then update the recipes, finally remove obsolete item definitions.

**Comment by mugling** [comment]
I'm going to split the migration code out first into a separate PR

**Comment by Coolthulhu** [comment]
Has a conflict, but it looks trivial. `git mergetool` seems to have fixed it.

File: data/json/items/classes/comestible.json [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    "symbol": "%"
+  },
+  {
+    "id": "fruit",
+    "copy-from": "food",
+    "type": "COMESTIBLE",
+    "name": "base fruit",
+    "material": "fruit",
+    "color": "red",
+    "vitamins": [ [ "VitC", 15 ] ]
```
Comment (Coolthulhu) [inline data/json/items/classes/comestible.json]: I think it's "vitC", not "VitC". A warning in finalization could be useful.

Comment (mugling) [inline data/json/items/classes/comestible.json]: Fixed in 1e1ef6c

**Comment by mugling** [comment]
@Coolthulhu fixed

**Comment by mugling** [comment]
Ready

**Comment by Coolthulhu** [comment]
Could you rebase this? Would make merging simpler

**Comment by mugling** [comment]
Yes, I'm waiting for the existing PR queue to die down somewhat first though

**Comment by mugling** [comment]
@Coolthulhu I've refactored #16694 which can be considered the first stages of this PR. Is this too late for you to try also as if not I'll subsequently rebase this PR as well and that in turn unblocks a few others

**Comment by mugling** [comment]
@Coolthulhu heavily rebased and ready for merge testing

File: data/json/items/classes/comestible.json [hunk 14]
```diff
@@ -0,0 +1,61 @@
+[
+  {
+    "id": "food",
```
Comment (kevingranade) [inline data/json/items/classes/comestible.json]: Setting comestible type and symbol is reasonable, but volume:1 is not a property of an edible object, there are a large number of counterexamples, and as I've told you Re: gun inheritance, it's not ok to spread out item stats across various levels like this.

File: data/json/items/classes/comestible.json [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    "volume": 1,
+    "symbol": "%"
+  },
+  {
+    "id": "fruit",
+    "copy-from": "food",
+    "type": "COMESTIBLE",
+    "name": "base fruit",
+    "material": "fruit",
+    "color": "red",
```
Comment (kevingranade) [inline data/json/items/classes/comestible.json]: Same with color here, it's not always red.  Though in this case I believe you've arbitrarily changed the color of many fruits to match your template.

File: data/json/items/classes/comestible.json [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    "vitamins": [ [ "vitC", 15 ] ]
+  },
+  {
+    "id": "fruit_berry",
+    "copy-from": "fruit",
+    "type": "COMESTIBLE",
+    "name": "fruit berries",
+    "stack_size": 8,
```
Comment (kevingranade) [inline data/json/items/classes/comestible.json]: Same as volume above, the stack size is not a property of berries, in fact they never had it before, why did you add it?

File: data/json/items/classes/comestible.json [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    "spoils_in": 72
```
Comment (kevingranade) [inline data/json/items/classes/comestible.json]: Spoil time is probably reasonable for all berries.

**Comment by kevingranade** [comment]
I'm trying to come up with an acceptable way to support what you're trying to do here without starting down the path of the jsons being impossible for humans to read.
The only thing that comes to mind is to not use the override mechanism, if you define a property of a class of items, it must be universal for that class of items.  In that case there is at least a consistent place to find every stat.

As I pointed out inline, volume, stack size and color are not properties of the various groups of items, they are just nearly consistent, and require either artificially homogenizing item definitions or adding a number of exceptions and making the definitions hard to read.

I recommend that you strip out everything but irradiated vegetable and vitamin definitions, in fact I recommend that you split them into one PR each as they have nothing to do with each other.

Also RE: irradiated vegetables, preventing their spawn is still unacceptable.

File: data/json/items/classes/comestible.json (hunk 14 above)
Comment (mugling) [inline data/json/items/classes/comestible.json]: This was changed in response to @Coolthulhu earlier review:

> Why explicitly specify 1 volume for everything?

It's possible we should specify `volume=1` as the default for `itype`?

File: data/json/items/classes/comestible.json (hunk 15 above)
Comment (mugling) [inline data/json/items/classes/comestible.json]: That's intentional. The idea being that different groupings of items should have the same colour and symbol?

File: data/json/items/classes/comestible.json (hunk 16 above)
Comment (mugling) [inline data/json/items/classes/comestible.json]: Again code comment from @Coolthulhu. I'm not especially averse to either proposal but this PR is going in circles.

**Comment by mugling** [comment]
> As I pointed out inline, volume, stack size and color are not properties of the various groups of items

These were in response to earlier code review comments. How do we go about resolving the difference of opinion in this respect?

File: data/json/items/classes/comestible.json (hunk 14 above)
Comment (Coolthulhu) [inline data/json/items/classes/comestible.json]: Maybe specify it as default for subtypes then? Vast majority of fruit is volume 1 or volume 1 and stack size 4 or 8.

Per-item volume 1 is an overkill for sets of nearly identical objects (ie. fruit) and per-category defaults are less ugly than hardcoded defaults.

This may change if the "volume to ml" PR lands (and obsoletes stack size), but until then, it's a reasonable assumption that fruit has 1 volume.

File: data/json/items/classes/comestible.json (hunk 16 above)
Comment (Coolthulhu) [inline data/json/items/classes/comestible.json]: IRL berries are quite small. Stack size 4-8 would represent their size better.

Alternatively, name change from "raspberry" to "handful of raspberries" to avoid having to change numbers in recipes.

File: data/json/items/classes/comestible.json (hunk 14 above)
Comment (mugling) [inline data/json/items/classes/comestible.json]: I'm broadly of the same opinion. Comments continue at the bottom of this thread

**Comment by mugling** [comment]
I don't think such changes reduce readability with the evidence being the recent increase in the number of new developers making JSON contributions using this format. Everyone agrees we need to attract and retain new developers.

The minority of new developers using the original approach of large files with fully explicit fields have made less progress and indeed some of their recent contributions have broken `master`. Those PR's are harder to review and more error-prone.

**Comment by kevingranade** [comment]
You're not reading my comments then, it is the path you are on I object to,
not the immediate effects.  This kind of hierarchy is a nightmare in code,
and it will be a nightmare in data if it continues, and I do not believe
that it will stop where it is, so I'm pushing back.

Each level of inheritance should have sole ownership of its data.
Each level of inheritance should own data that is definitional for its name.
The levels should be small and meaningful in number.

This is just simple OO principles, because that is exactly what this us.

Inheritance is NOT for "reducing redundancy" it is for properly modelling
objects in an easy to maintain way. Reducing redundancy is a nice side
effect.

**Comment by mugling** [comment]
Reimplementation would likely be faster than rebase and the `requirements` fragment system partially mitigates some of the problems that originally motivated this PR
